Add all supported languages to Google Translate and remove unsupported ones

prabhjotsbhatia-ca opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->


## Proposed change
This PR proposes to add supported languages from Google Translate. 
https://www.home-assistant.io/integrations/google_translate/

Earlier, not all languages were showing up, even though the documentation stated they were supported. I was personally missing a single language, but added all of the missing ones so that they don't need to be requested again. 

I have manually verified that the following supported languages were missing: 

- am
- eu
- fil
- gl
- ha
- ms
- pa

In addition, the following unsupported languages were added in Home Assistant: (These silently fail right now) 
- eo
- hy
- mk
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #107403


## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [ ] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

prabhjotsbhatia-ca committed:
Adding supported language codes from Google Translate

Added all languages that Google Translate supports.

prabhjotsbhatia-ca committed:
Corrected alphabetical order of languages

prabhjotsbhatia-ca committed:
Merge branch 'dev' into dev

prabhjotsbhatia-ca committed:
Merge branch 'dev' into dev

prabhjotsbhatia-ca committed:
Merge branch 'dev' into dev

prabhjotsbhatia-ca committed:
Merge branch 'dev' into dev

prabhjotsbhatia-ca committed:
Merge branch 'dev' into dev

prabhjotsbhatia-ca committed:
Merge branch 'dev' into dev

prabhjotsbhatia-ca committed:
Merge branch 'dev' into dev

prabhjotsbhatia-ca committed:
Merge branch 'dev' into dev

emontnemery reviewed (approved):
LGTM, thanks @prabhjotsbhatia-ca 👍

emontnemery reviewed (commented):
How did you come up with the list of languages supported for text to speech?
I find this list https://cloud.google.com/text-to-speech/docs/voices which seems to be reasonably well aligned with the currently supported languages, although we do miss supported languages.

prabhjotsbhatia-ca commented:
> How did you come up with the list of languages supported for text to speech? I find this list https://cloud.google.com/text-to-speech/docs/voices which seems to be reasonably well aligned with the currently supported languages, although we do miss supported languages.

I used the link on the Home assistant documentation https://www.home-assistant.io/integrations/google_translate/
which points to https://cloud.google.com/translate/docs/languages

emontnemery commented:
> > How did you come up with the list of languages supported for text to speech? I find this list https://cloud.google.com/text-to-speech/docs/voices which seems to be reasonably well aligned with the currently supported languages, although we do miss supported languages.
> 
> I used the link on the Home assistant documentation https://www.home-assistant.io/integrations/google_translate/ which points to https://cloud.google.com/translate/docs/languages

Right, but I wonder if that's the correct link. You mentioned there's a specific language you want to add, where does your language appear? Have you verified that tts works for your language as well as for other languages you've added?

prabhjotsbhatia-ca commented:
I wanted to use pa (Punjabi) but wasn't able to even though the TLD (com.pa) is added in Home assistant.
The other language codes from the list on https://cloud.google.com/translate/docs/languages work perfectly , but I didn't test with the ones in tts docs -  the first has codes in lowercase, "en-us", "en-gb" etc whereas the latter  has "en-US" ën-GB" etc. 
(I've tested lowercase works)
I believe the Google Translate API and Cloud TTS are also distinct from each other.

emontnemery commented:
I see, Punjabi is listed on https://cloud.google.com/text-to-speech/docs/voices however.

If you try a handful of languages added in your PR, but not listed on https://cloud.google.com/text-to-speech/docs/voices, do they too work?

prabhjotsbhatia-ca commented:
Yes, I just finished testing this with tl (Tagalog) which doesn't exist on
https://cloud.google.com/text-to-speech/docs/voices, but exists in home
assistant right now. The equivalent code in cloud TTS is likely fil-PH".
- Created a new Google translate integration  entity with tl language and
tl TLD.
- Then passed on a Tagalog sentence to it, and called the service.
- Google assistant spoke in what sounded like a Tagalog speaker.


My understanding is that the Google Translate API is not billed unlike
cloud TTS which is billable to your Google account.

On Wed, Apr 17, 2024, 1:29 PM Erik Montnemery ***@***.***>
wrote:

> I see, Punjabi is listed on
> https://cloud.google.com/text-to-speech/docs/voices however.
>
> If you try a handful of languages added in your PR, but not listed on
> https://cloud.google.com/text-to-speech/docs/voices, do they too work?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/home-assistant/core/pull/107404#issuecomment-2061832841>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ANQ66IHPJAVOH25WHII2RS3Y52WPLAVCNFSM6AAAAABBPXS6V2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDANRRHAZTEOBUGE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

emontnemery commented:
Right, but I think we need to come up with a better way to validate the list of languages which support tts via the Google Translate API.

Starting from the top, your PR adds the "Akan" language, which based on the web UI within Google search can't be spoken:
![image](https://github.com/home-assistant/core/assets/14281572/1d7b74ba-273d-41f5-9054-c917830da338)

It's not even possible to pick on translate.google.com
![image](https://github.com/home-assistant/core/assets/14281572/7e3a71df-b294-43d9-b411-0ad69d5b2e3e)

prabhjotsbhatia-ca committed:
Merge branch 'home-assistant:dev' into dev

prabhjotsbhatia-ca commented:
> Right, but I think we need to come up with a better way to validate the list of languages which support tts via the Google Translate API.
> 
> Starting from the top, your PR adds the "Akan" language, which based on the web UI within Google search can't be spoken: ![image](https://private-user-images.githubusercontent.com/14281572/323487769-1d7b74ba-273d-41f5-9054-c917830da338.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTM0OTczOTYsIm5iZiI6MTcxMzQ5NzA5NiwicGF0aCI6Ii8xNDI4MTU3Mi8zMjM0ODc3NjktMWQ3Yjc0YmEtMjczZC00MWY1LTkwNTQtYzkxNzgzMGRhMzM4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDE5VDAzMjQ1NlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTM3NDE2YTgyOWZjZWQzODIyYjFlYmIwNmUxZjE0MTU4Y2NlZjkwYWEzNWMyYWZmYjVmN2MxODI0Y2E0ZTk1NWUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.MLqax90-aJtaYCviXlb1ChyW1YKDOM88jK45Bci5XUc)
> 
> It's not even possible to pick on translate.google.com ![image](https://private-user-images.githubusercontent.com/14281572/323488138-7e3a71df-b294-43d9-b411-0ad69d5b2e3e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTM0OTczOTYsIm5iZiI6MTcxMzQ5NzA5NiwicGF0aCI6Ii8xNDI4MTU3Mi8zMjM0ODgxMzgtN2UzYTcxZGYtYjI5NC00M2Q5LWI0MTEtMGFkNjlkNWIyZTNlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDE5VDAzMjQ1NlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTUxMWM4NDE1YzJhOGMxNjY2YWI4MGYxMzAyYjVmNjAxOWJiMmFlMjVmYjFjNmI5NmFlYjZiNDA3ZDA0MmM2ZTYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.NAr-ujpsBttsNT6HaKMvKtMDjv7UxAnfXIEWKPZyeXU)

You're absolutely right. Though I can't find an authoritative list which has all the languages available via Google Translate Speech. The Cloud TTS page linked has several missing; there are several languages not present there but which have speech available via Google Translate. 
I am manually compiling a list from Google translate as to which languages have speech available.

prabhjotsbhatia-ca committed:
Remove languages not actually supported for speech

Previously I added languages supported by Google Translate. Based on comments received, I manually verified each language, and  removed languages that are not actually supported for speech in Google Translate.

prabhjotsbhatia-ca commented:
Finished manual verification. Only a handful of languages actually supported by speech in Google Translate were missing. In addition, I found three languages which existed in Home Assistant but are not actually supported by Google Translate. 
I have verified these don't work in home assistant right now (silently fails).

emontnemery commented:
> Finished manual verification. Only a handful of languages actually supported by speech in Google Translate were missing. In addition, I found three languages which existed in Home Assistant but are not actually supported by Google Translate. I have verified these don't work in home assistant right now (silently fails).

Thanks a lot @prabhjotsbhatia-ca 👍

To help ourselves the next time someone wants to change this, could you add a comment above the `SUPPORT_LANGUAGES` (briefly) explaining how to update the list?

I also think we should change the documentation, maybe it would be better to just remove the links to Google since what we support is fixed in the code; linking to some ever changing Google documentation is not helpful.

prabhjotsbhatia-ca committed:
Add instructions to update the list of supported languages

Added instructions as suggested so as to facilitate easier update on this list.

prabhjotsbhatia-ca commented:
@emontnemery  I've added those instructions to the code. Not sure on the best way to go forward on the Home Assistant documentation. Could you pleaes help?

emontnemery commented:
> Not sure on the best way to go forward on the Home Assistant documentation. Could you pleaes help?

I meant that I think it would be better to replace the misleading text here https://www.home-assistant.io/integrations/google_translate/#configuration with a table which lists the languages supported in the code.

emontnemery committed:
Reformat comment in const.py
